functional/stateful-types.md

```diff
@@ -0,0 +1,269 @@
+constraint", and each value used by an end user *actually changes the type of
+each instatiation*. In other words, `Vec<usize>` and `Vec<char>` *are two
+different types*.
+
+This is why `impl` blocks must specify generic parameters: different ones can
+have different `impl` blocks on them.
+
+It is recommended in rust to use generic types to enforce invariants. The best
+example is a state machine.
+
+## Example
+
+Suppose you are designing an interpreted language runtime, which requires a JIT
+compiler in order to process. Many Rust novices coming from other languages
+would implement it this way:
+
+```rust
+#[derive(Debug)]
+pub enum CompileError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+#[derive(Debug)]
+pub enum ExecError {
+    // error type, std::error::Error impl skipped for brevity
+}
+
+pub type ExecutionResult = Result<(), ExecError>;
+
+pub type CompileResult = Result<(), CompileError>;
+
+pub enum Param {
+    GCSize(usize),
+    MaxStack(usize),
+    // more parameters...
+}
+
+#[derive(Default)]
+pub struct Interpreter {
+    // state and execution data
+}
+
+impl Interpreter {
+    /// Set a system prameter.
+    ///
+    /// # Panics
+    /// Will panic if called after a script is executed.
+    ///
+    pub fn set_param(&mut self, p: Param) {
+        /* update the state based on the parameter */
+    }
+
+    /// Set up a new execution environment.
+    ///
+    /// # Panics
+    /// Will panic if called more than once or after a script has been compiled.
+    ///
+    pub fn init(&mut self) {
+        /* Create the heap, prepare for use... */
+    }
+
+    /// Load and compile a script. Returns any errors encountered.
+    ///
+    pub fn compile_script(&mut self, script: &str) -> CompileResult {
+        /* actually do the compile... */
+        Ok(())
+    }
+
+    /// Execute all scripts added. Returns any errors encountered.
+    ///
+    /// # Panics
+    /// Will panic if zero scripts have been compiled.
+    ///
+    pub fn exec(&mut self) -> ExecutionResult {
+        /* actually run the script... */
+        Ok(())
+    }
+}
+
+fn main() {
+    let mut interp = Interpreter::default();
+    interp.set_param(Param::GCSize(1024 * 1024 * 1024));
+    interp.init();
+    interp.compile_script("print('2 + 2')").unwrap();
+    interp.exec().unwrap();
+}
+```
+
+Why those chances to panic? Because there are *state invariants* here:
+
+1. `set_param` can only be called in an "initial" state.
+2. `init` must be called before any scripts are compiled.
+3. `exec` can only be called in a "loaded" or "ready" state.
+4. `init` must be called *exactly once*.
+
+It would be possible to add these to the `Error` types instead of panicking.
+However, this solution is suboptimal. Not only would users of the code
+correctly have to `unwrap()` a result all the time, but the requirement to call
+init exactly once is still unenforcable across an entire program without
```

Comment:
```suggestion
`init()` exactly once is still unenforceable across an entire program without
```

Reply:
Should we enable spell checkers?

Reply:
pickfire, you could open an issue about this. It doesn't seem like a bad idea. But we will need to have our own dictionary. By the way, let's continue the discussion in a proper issue, not here.